In [22]:
from sqlalchemy import create_engine
import psycopg2 as db
import pandas as pd
import numpy as np
import os

import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
pd.options.display.max_rows = 999
pd.options.display.max_columns= 999

In [3]:
postPass=os.environ["POSTGRES_PASS"]

In [4]:
engine = create_engine("postgresql+psycopg2://{user}:{pw}@localhost/{db}"
                       .format(user="jupyter", pw=postPass, db="expunge"))

# Table 1: Metrics by FIPS code, code section, and race
One row for every code_section/fips/race, with columns for count within race, count overall, proportion, and disparity index. At the individual level

## a. individual level these results do not make sense

In [ ]:
table_1a = """
WITH byrace AS (
    SELECT fips,race, code_section, COUNT(DISTINCT(person_id)) AS distinct_count_race 
    FROM charges
    GROUP BY fips, race, code_section),
overall AS (
    SELECT fips, code_section, COUNT(DISTINCT(person_id)) AS distinct_count_overall
    FROM charges
    GROUP BY fips, code_section)
SELECT c.race, c.code_section,c.fips, c.distinct_count_race , d.distinct_count_overall, l.proportion,
    (cast(c.distinct_count_race as decimal)/cast(d.distinct_count_overall as decimal))/cast(l.proportion as decimal) as disparity_index
FROM byrace c
INNER JOIN overall d
    ON c.code_section=d.code_section
INNER JOIN census_long l
    ON l.race=c.race
WHERE distinct_count_overall > 1000
ORDER BY disparity_index DESC
"""
table1a_ = pd.read_sql_query(table_1a, con=engine)

In [ ]:
table1a_

In [17]:
table1a = """
WITH byrace AS (
    SELECT fips, race, code_section, COUNT(DISTINCT(person_id)) AS distinct_count_race 
    FROM charges
    GROUP BY fips, race, code_section),
overall AS (
    SELECT fips, code_section, COUNT(DISTINCT(person_id)) AS distinct_count_overall 
    FROM charges
    GROUP BY fips, code_section)
SELECT c.fips, l."Jurisdiction" as jurisdiction, c.race, c.code_section, c.distinct_count_race, d.distinct_count_overall, l.proportion,
    (cast(c.distinct_count_race as decimal)/cast(d.distinct_count_overall as decimal))/cast(l.proportion as decimal) as disparity_index
FROM byrace c
INNER JOIN overall d
    ON c.fips = d.fips AND c.code_section=d.code_section
INNER JOIN census_long l
    ON cast(c.fips as int) = l."FIPS" AND l.race=c.race
WHERE distinct_count_overall > 1000    
ORDER BY disparity_index DESC
"""
table_1a = pd.read_sql_query(table1a, con=engine)

In [18]:
table_1a

fips          jurisdiction      race code_section  distinct_count_race  \
0     041   Chesterfield County   Unknown        13-60                 1807   
1     081    Greensville County   Unknown     46.2-300                  764   
2     035        Carroll County   Unknown   A.46.2-862                  837   
3     540  Charlottesville city   Unknown     18.2-181                  224   
4     113        Madison County   Unknown   A.46.2-862                  163   
...   ...                   ...       ...          ...                  ...   
6575  059        Fairfax County  Hispanic  18.2-108.01                    2   
6576  059        Fairfax County  Hispanic     18.2-178                    2   
6577  013      Arlington County  Hispanic     18.2-388                    1   
6578  059        Fairfax County  Hispanic     18.2-195                    1   
6579  059        Fairfax County  Hispanic     18.2-192                    1   

      distinct_count_overall  proportion  disparity_index  
0                       2348    0.007208       106.768893  
1                       4269    0.002911        61.477062  
2                       4686    0.003793        47.090122  
3                       1081    0.004697        44.118242  
4                       1237    0.003016        43.685186  
...                      ...         ...              ...  
6575                    2197    0.165150         0.005512  
6576                    2214    0.165150         0.005470  
6577                    1237    0.156015         0.005182  
6578                    1180    0.165150         0.005131  
6579                    1653    0.165150         0.003663  

[6580 rows x 8 columns]

In [21]:
table_1a.query("race=='Black'")

fips          jurisdiction   race code_section  distinct_count_race  \
24    035        Carroll County  Black     46.2-300                  988   
27    035        Carroll County  Black   C.46.2-862                 3001   
28    035        Carroll County  Black   18.2-250.1                  248   
48    035        Carroll County  Black   A.46.2-862                  670   
86    035        Carroll County  Black   B.46.2-301                  221   
...   ...                   ...    ...          ...                  ...   
4604  147  Prince Edward County  Black      4.1-305                  148   
4825  059        Fairfax County  Black       36-106                   28   
4924  105            Lee County  Black   B.46.2-301                   13   
5063  185       Tazewell County  Black     46.2-752                    9   
5659  027       Buchanan County  Black   B.46.2-301                    5   

      distinct_count_overall  proportion  disparity_index  
24                      4773    0.008828        23.447408  
27                     15712    0.008828        21.635347  
28                      1359    0.008828        20.671003  
48                      4686    0.008828        16.195779  
86                      2010    0.008828        12.454479  
...                      ...         ...              ...  
4604                    1515    0.328392         0.297479  
4825                    1080    0.106275         0.243951  
4924                    1544    0.037228         0.226163  
5063                    1418    0.032270         0.196683  
5659                    1333    0.033803         0.110964  

[1114 rows x 8 columns]

## b. charge level

In [ ]:
table1b = """
WITH byrace AS (
    SELECT fips, race, code_section, COUNT(*) AS total_count_race 
    FROM charges
    GROUP BY fips, race, code_section),
overall AS (
    SELECT fips, code_section, COUNT(*) AS total_count_overall
    FROM charges
    GROUP BY fips, code_section)
SELECT c.fips, c.race, c.code_section,  c.total_count_race, d.total_count_overall, l.proportion,
    (cast(c.total_count_race as decimal)/cast(d.total_count_overall as decimal))/cast(l.proportion as decimal) as disparity_index
FROM byrace c
INNER JOIN overall d
    ON c.code_section=d.code_section
INNER JOIN census_long l
    ON l.race=c.race
WHERE total_count_overall > 1000
ORDER BY disparity_index DESC
"""
table1b_ = pd.read_sql_query(table1b, con=engine)

In [9]:
myquery = """
WITH byrace AS (
    SELECT fips, race, code_section, COUNT(*) AS total_count_race 
    FROM charges
    GROUP BY fips, race, code_section),
overall AS (
    SELECT fips, code_section, COUNT(*) AS total_count_overall 
    FROM charges
    GROUP BY fips, code_section)
SELECT c.fips, l."Jurisdiction" as jurisdiction, c.race, c.code_section, c.total_count_race, d.total_count_overall, l.proportion,
    (cast(c.total_count_race as decimal)/cast(d.total_count_overall as decimal))/cast(l.proportion as decimal) as disparity_index
FROM byrace c
INNER JOIN overall d
    ON c.fips = d.fips AND c.code_section=d.code_section
INNER JOIN census_long l
    ON cast(c.fips as int) = l."FIPS" AND l.race=c.race
WHERE total_count_overall > 1000    
ORDER BY disparity_index DESC
"""
disp_index = pd.read_sql_query(myquery, con=engine)

In [10]:
disp_index

fips           jurisdiction      race code_section  total_count_race  \
0     041    Chesterfield County   Unknown        13-60              1990   
1     081     Greensville County   Unknown     46.2-300               769   
2     035         Carroll County   Unknown   A.46.2-862               841   
3     113         Madison County   Unknown   A.46.2-862               164   
4     081     Greensville County   Unknown   A.46.2-862             12814   
...   ...                    ...       ...          ...               ...   
8564  013       Arlington County  Hispanic     18.2-388                 1   
8565  059         Fairfax County  Hispanic  18.2-108.01                 2   
8566  153  Prince William County  Hispanic     18.2-192                 1   
8567  059         Fairfax County  Hispanic     18.2-178                 2   
8568  153  Prince William County  Hispanic   18.2-258.1                 1   

      total_count_overall  proportion  disparity_index  
0                    2638    0.007208       104.655718  
1                    4345    0.002911        60.797043  
2                    4713    0.003793        47.044103  
3                    1256    0.003016        43.288296  
4                  102739    0.002911        42.844520  
...                   ...         ...              ...  
8564                 1437    0.156015         0.004460  
8565                 2969    0.165150         0.004079  
8566                 1114    0.244874         0.003666  
8567                 3512    0.165150         0.003448  
8568                 1231    0.244874         0.003317  

[8569 rows x 8 columns]

# Table 2: Metrics by code section and race
One row for every code_section/race, with columns for count within race, count overall, proportion, and disparity index. At the individual level

## a. individual level (for distinct_count_overall > 1000)

In [42]:
table_2a = """
WITH byrace AS (
    SELECT race, code_section, COUNT(DISTINCT(person_id)) AS distinct_count_race 
    FROM charges
    GROUP BY race, code_section),
overall AS (
    SELECT code_section, COUNT(DISTINCT(person_id)) AS distinct_count_overall
    FROM charges
    GROUP BY code_section)
SELECT c.race, c.code_section, m.max_fips, c.distinct_count_race , d.distinct_count_overall, l.proportion,
    (cast(c.distinct_count_race as decimal)/cast(d.distinct_count_overall as decimal))/cast(l.proportion as decimal) as disparity_index
FROM byrace c
INNER JOIN overall d
    ON c.code_section=d.code_section
INNER JOIN census_va l
    ON l.race=c.race
INNER JOIN maxfips m 
    ON m.code_section = c.code_section
WHERE distinct_count_overall > 1000
ORDER BY disparity_index DESC
"""
table2a = pd.read_sql_query(table_2a, con=engine)

In [43]:
table2a

race  code_section max_fips  distinct_count_race  \
0                       Unknown         13-60      041                 1807   
1                       Unknown        16-3.1      810                  261   
2                         Black        24-253      740                 1648   
3                         Black         29-48      711                 3585   
4                         Black      60.2-632      761                 2006   
...                         ...           ...      ...                  ...   
1772  Asian or Pacific Islander        24-253      740                    1   
1773                   Hispanic  18.2-308.1:4      810                    1   
1774                   Hispanic          29-8      711                    2   
1775                   Hispanic        46-157      550                    2   
1776                   Hispanic  19.2-152.4:1      153                    1   

      distinct_count_overall  proportion  disparity_index  
0                       2348    0.038223        20.134436  
1                       1462    0.038223         4.670599  
2                       1936    0.198806         4.281764  
3                       4382    0.198806         4.115169  
4                       2519    0.198806         4.005656  
...                      ...         ...              ...  
1772                    1936    0.069089         0.007476  
1773                    1412    0.097759         0.007245  
1774                    2834    0.097759         0.007219  
1775                    3401    0.097759         0.006015  
1776                    2066    0.097759         0.004951  

[1777 rows x 7 columns]

## b. charge level (where total_count_overall > 1000)

In [44]:
table_2b = """
WITH byrace AS (
    SELECT race, code_section, COUNT(*) AS total_count_race 
    FROM charges
    GROUP BY race, code_section),
overall AS (
    SELECT code_section, COUNT(*) AS total_count_overall
    FROM charges
    GROUP BY  code_section)
SELECT c.race, c.code_section, m.max_fips, c.total_count_race, d.total_count_overall, l.proportion,
    (cast(c.total_count_race as decimal)/cast(d.total_count_overall as decimal))/cast(l.proportion as decimal) as disparity_index
FROM byrace c
INNER JOIN overall d
    ON c.code_section=d.code_section
INNER JOIN census_va l
    ON l.race=c.race
INNER JOIN maxfips m 
    ON m.code_section = c.code_section
WHERE total_count_overall > 1000
ORDER BY disparity_index DESC
"""
table2b = pd.read_sql_query(table_2b, con=engine)

In [45]:
table2b

race  code_section max_fips  total_count_race  \
0                       Unknown         13-60      041              1990   
1                         Black        24-253      740              3372   
2                       Unknown        16-3.1      810               482   
3                         Black         29-48      711              6106   
4                         Black     18.2-53.1      760             42793   
...                         ...           ...      ...               ...   
2008                   Hispanic         24-13      650                 3   
2009                   Hispanic        46-157      550                 2   
2010  Asian or Pacific Islander        24-253      740                 1   
2011                   Hispanic  19.2-152.4:1      153                 1   
2012                   Hispanic       4.1-322      770                 2   

      total_count_overall  proportion  disparity_index  
0                    2638    0.038223        19.735935  
1                    3818    0.198806         4.442450  
2                    2925    0.038223         4.311224  
3                    7170    0.198806         4.283596  
4                   51333    0.198806         4.193213  
...                   ...         ...              ...  
2008                 6687    0.097759         0.004589  
2009                 4761    0.097759         0.004297  
2010                 3818    0.069089         0.003791  
2011                 3209    0.097759         0.003188  
2012                 7557    0.097759         0.002707  

[2013 rows x 7 columns]

# Table 3: Metrics by FIPS code and race
One row for every fips/race, with columns for count within race, count overall, proportion, and disparity index. At the individual level

## a. individual level

In [11]:
table_3a = """
WITH byrace AS (
    SELECT fips,race, COUNT(DISTINCT(person_id)) AS distinct_count_race 
    FROM charges
    GROUP BY fips, race),
overall AS (
    SELECT fips, COUNT(DISTINCT(person_id)) AS distinct_count_overall
    FROM charges
    GROUP BY fips)
SELECT c.fips, c.race, c.distinct_count_race , d.distinct_count_overall, l.proportion,
    (cast(c.distinct_count_race as decimal)/cast(d.distinct_count_overall as decimal))/cast(l.proportion as decimal) as disparity_index
FROM byrace c
INNER JOIN overall d
    ON c.fips = d.fips
INNER JOIN census_long l
    ON l.race=c.race
WHERE distinct_count_overall > 1000
ORDER BY disparity_index DESC
"""
table3a = pd.read_sql_query(table_3a, con=engine)

In [40]:
table_3a_test1 = """
WITH byrace AS (
    SELECT fips,race, COUNT(DISTINCT(person_id)) AS distinct_count_race 
    FROM charges
    GROUP BY fips, race),
overall AS (
    SELECT fips, COUNT(DISTINCT(person_id)) AS distinct_count_overall
    FROM charges
    GROUP BY fips)
SELECT c.fips, c.race, c.distinct_count_race , d.distinct_count_overall, l.proportion,
    (cast(c.distinct_count_race as decimal)/cast(d.distinct_count_overall as decimal))/cast(l.proportion as decimal) as disparity_index
FROM byrace c
INNER JOIN overall d
    ON c.fips = d.fips
INNER JOIN census_long l
    ON l.race=c.race
WHERE distinct_count_overall > 1000 AND d.fips = '760'
ORDER BY disparity_index DESC
"""
table3a_test1 = pd.read_sql_query(table_3a_test1, con=engine)

In [41]:
table3a_test1.shape

(798, 6)

Our census_long table has 798 rows. For one FIPS code in table3a_test1 there are 798 rows. It is probably not a coincidence that these row values are the same. 

In [36]:
table3a_test1.shape

(798, 6)

In [20]:
table3a.query("fips == '760'").shape

(798, 6)

## b. charge level